### ISSUE TITLE
Memory explosion when jitting dynamic_slice

### ISSUE THREAD

**Issue by Dapid** [google/jax#824]
The [following example][1] explodes in memory usage when calling the jitted functions. The culprit seems to be the dynamic_slice. It happens calling with or without the GPU.

Incidentally, it is also rather slow (twice slower than the plain Python version), so I welcome comments if there is a jax-friendlier way to rewrite it, but perhaps that is what jit should do?

My system:

Fedora 29, 31 GB RAM
jax 0.1.36
jaxlib 0.1.16, installed from the googleapis wheel.
CUDA 10.0
Nvidia driver 430.14.

   [1]: https://gist.github.com/Dapid/b33919674f6a3cd9839473bfd2821e28

**Comment by hawkinsp** [comment]
Sorry for the long delay in response, most of us were away last week.

I think the memory blow-up happens because the gradient of `dynamic_slice` constructs a dense array of zeros to scatter a value into, and we end up with one such constant for each `dynamic_slice` call. This is similar to #850 ; in general JAX probably needs to get smarter about not instantiating large constants.

However, there's a better way to write your computation to use a single gather (via NumPy advanced indexing) rather than a long chain of `dynamic_slice` calls. The JIT-ted version no longer blows up in memory and is much faster:

```
    def single_loss(self, datum):
        J = self.J
        N = self.N

        loss = 0.
        for r in range(N):
            x = np.sum(J[np.arange(N), r, datum[np.arange(N)], datum[r]])
            loss -= np.sum(np.exp(x))
        return loss
```

If you want to get even fancier you use a single gather for the entire outer loop too:

```
        xs = np.arange(N)
        ys = np.sum(J[xs[:, None], xs, datum[xs, None], datum[xs]], axis=1)
        loss = -np.sum(np.exp(ys))
```

(The second version didn't seem faster to me but it's fun to code-golf these things!)

**Comment by hawkinsp** [comment]
To add to my earlier comment: the issue I mentioned was one problem but not the only one. It turned out we were also duplicating array `J`. PR #873 fixes the memory blow up, but you should still prefer a single gather to a chain of dynamic slices (and NumPy indexing syntax is easier to read, anyway...)

**Comment by Dapid** [comment]
Amazing! I am getting three orders of magnitude speed-up. Thank you for jax!

I have noticed that taking the grad of my new function needs a lot of memory, but `grad(jit(f))` works fine, and `jit(grad(jit(f)))` even better. Is this something I am supposed to do? Are there general guidelines, or just experimentation?

**Comment by hawkinsp** [comment]
In general placing a `jit` as the outermost transformation gives the XLA compiler the most opportunities for optimization (e.g., fusing broadcasts into their consumers).

That said, if you have an example of particularly bad memory blow up outside a jit, we could certainly look into it.
